sharding/contracts/validator_manager.sol

```diff
@@ -0,0 +1,229 @@
+    require(block.number >= periodLength);
+    require(_expectedPeriodNumber == block.number / periodLength);
+    require(_periodStartPrevHash == block.blockhash(_expectedPeriodNumber * periodLength - 1));
+
+    // Check if this header already exists
+    headerVars.entireHeaderHash = keccak256(_shardId, _expectedPeriodNumber, _periodStartPrevHash,
+                                   _parentCollationHash, _txListRoot, bytes32(_collationCoinbase),
+                                   _postStateRoot, _receiptRoot, _collationNumber);
+    assert(headerVars.entireHeaderHash != 0x0);
+    assert(collationHeaders[_shardId][headerVars.entireHeaderHash].score == 0);
+    // Check whether the parent exists.
+    // if (parent_collation_hash == 0), i.e., is the genesis,
+    // then there is no need to check.
+    if (_parentCollationHash != 0x0)
+        assert((_parentCollationHash == 0x0) || (collationHeaders[_shardId][_parentCollationHash].score > 0));
+    // Check if only one collation in one period
+    assert(periodHead[_shardId] < int(_expectedPeriodNumber));
+
+    // Check the signature with validation_code_addr
+    headerVars.validatorAddr = getEligibleProposer(_shardId, block.number/periodLength);
+    require(headerVars.validatorAddr != 0x0);
+    require(msg.sender == headerVars.validatorAddr);
+
+    // Check score == collationNumber
+    headerVars.score = collationHeaders[_shardId][_parentCollationHash].score + 1;
+    require(_collationNumber == headerVars.score);
+
+    // Add the header
+    collationHeaders[_shardId][headerVars.entireHeaderHash] = CollationHeader({
+      parentCollationHash: _parentCollationHash,
+      score: headerVars.score
+    });
+
+    // Update the latest period number
+    periodHead[_shardId] = int(_expectedPeriodNumber);
+
+    // Determine the head
+    if (headerVars.score > collationHeaders[_shardId][shardHead[_shardId]].score) {
+      shardHead[_shardId] = headerVars.entireHeaderHash;
+      headerVars.isNewHead = true;
+    }
+    // [TODO] Log
+    //CollationAdded(headerBytes, isNewHead, _score);
+
+    return true;
+  }
+
+  // Returns the gas limit that collations can currently have (by default make
+  // this function always answer 10 million).
+  function getCollationGasLimit() public pure returns(uint) {
+    return 10000000;
+  }
+
+  // Records a request to deposit msg.value ETH to address to in shard shard_id
+  // during a future collation. Saves a `receipt ID` for this request,
+  // also saving `msg.sender`, `msg.value`, `to`, `shard_id`, `startgas`,
+  // `gasprice`, and `data`.
+  function txToShard(address _to, int _shardId, uint _txStartgas, uint _txGasprice, bytes12 _data) public payable returns(int) {
+    receipts[numReceipts] = Receipt({
+      shardId: _shardId,
+      txStartgas: _txStartgas,
+      txGasprice: _txGasprice,
+      value: msg.value,
+      sender: msg.sender,
+      to: _to,
+      data: _data
+    });
+    var receiptId = numReceipts;
+    ++numReceipts;
```

Comment:
I'm not 100% on solidity, but if receiptId is the same reference as numReceipts then wouldn't they both be incremented?

Reply:
No, receiptId is a variable in memory.

Reply:
Sounds good